grails-app/services/org/pih/warehouse/core/LocationService.groovy

```diff
@@ -718,4 +724,83 @@ class LocationService {
+        location.organization = params.organization ? Organization.findByCodeOrName(params.organization, params.organization) : null
+        location.address = createOrUpdateAddress(params, location?.address?.id)
+        if (!location.inventory && !location.parentLocation) {
+            location.inventory = new Inventory(['warehouse': location])
+        }
+        def locationType = null
+        if (params.locationType) {
+            String locationTypeName = LocalizationUtil.getDefaultString(params.locationType as String)
+            // TODO: Replace with a single GORM .find with Closure when in Grails 3 (available since Grails 2.0)
+            LocationType matchedLocationType = LocationType
+                    .findAllByNameLike(locationTypeName + "%")
+                    .find{ LocalizationUtil.getDefaultString(it.name) == locationTypeName}
+            locationType = matchedLocationType
+        }
+
+        def currentLocationType = location.locationType
+
+        //Do not allow to change internal type location to non-internal
+        if (((currentLocationType?.isInternalLocation() || currentLocationType?.isZone()) && (locationType?.isInternalLocation() || locationType?.isZone())) || !(currentLocationType?.isInternalLocation() || currentLocationType?.isZone())) {
+            location.locationType = locationType
+            location.supportedActivities = currentLocationType && locationType == currentLocationType ? location.supportedActivities : location.supportedActivities?.clear()
+        }
+
+        // Add required association to organization for depots and suppliers
+        if (!(location.locationType?.isInternalLocation() || location.locationType?.isZone()) && !location.organization) {
+            def locationCode = identifierService.generateOrganizationIdentifier(params.organization)
+            Organization organization = (location.locationType?.locationTypeCode == LocationTypeCode.SUPPLIER) ?
+                    organizationService.findOrCreateSupplierOrganization(params?.organization, locationCode) :
+                    organizationService.findOrCreateOrganization(params?.organization, locationCode)
+
+            location.organization = organization
+        }
+
+        return location
+    }
+
+    Address createOrUpdateAddress(Map params, @Nullable String addressId) {
+        Address address
+        if (addressId) {
+            address = Address.findById(addressId)
+        }
+
+        if (!address) {
+            address = new Address()
+        }
+
+        address.address = params.streetAddress ?: ''
+        address.address2 = params.streetAddress2 ?: ''
+        address.city = params.city ?: ''
+        address.stateOrProvince = params.stateOrProvince ?: ''
+        address.postalCode = params.postalCode ?: ''
+        address.country = params.country ?: ''
+        address.description = params.description ?: ''
+
+        if (address.validate() && !address.hasErrors()) {
```

Comment:
I think `address.validate()` should be enough, since if it doesn't pass the validation, it'd return `false`, hence it wouldn't proceed with the `address.save()`

Reply:
Going back to @kchelstowski's point about standardizing, perhaps we just need a method name that reflect what it's doing.
```
Address address = buildAddress(params)
Address address = populateAddress(params)
```
Or actually, it's possible doing more of a findOrCreate
```
Address address = findOrCreateAddress(params)
```